src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1721 @@
+                c (250 bits) = bits 4..=253 of pk★_d
+                d (60 bits)  = d_0 || d_1 || d_2 || d_3
+                            = (bit 254 of x(pk_d)) || (ỹ bit of pk_d) || (0..=7 of v) || (8..=57 of v)
+                e (10 bits)  = e_0 || e_1
+                            = (bits 58..=63 of v) || (bits 0..=3 of rho)
+                f (250 bits) = bits 4..=253 inclusive of rho
+                g (250 bits) = g_0 || g_1 || g_2
+                            = (bit 254 of rho) || (bits 0..=8 of psi) || (bits 9..=248 of psi)
+                h (10 bits)  = h_0 || h_1 || h_2
+                            = (bits 249..=253 of psi) || (bit 254 of psi) || 4 zero bits
+
+                |         A_0         |          A_1           |         A_2            |          A_3            |  A_4  |  A_5  |  A_6  |  A_7   |  A_8  |   A_9   |  q_canon_1  |  q_canon_2  |
+                ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+                |       a_prime       |       b3_c_prime       |       e1_f_prime       |       g1_g2_prime       |   a   |   b   |  b_2  |   b_3  |   c   |    d    |      0      |      0      |
+                |         d_1         |           d_2          |          z1_d          |           e             |  e_0  |  e_1  |   f   |    g   |  g_1  |    h    |      1      |      0      |
+                |         h_0         |           h_1          |         x(g_d)         |        x(pk_d)          | value |  b_0  |  b_1  |   d_0  |  g_0  |   z1_g  |      0      |      1      |
+                |a_prime_decomposition|b3_c_prime_decomposition|e1_f_prime_decomposition|g1_g2_prime_decomposition| z13_a | z13_c | z13_f | z13_g  |  psi  |   rho   |      0      |      0      |
+            */
+
+            // q_canon_2 checks that:
+            //   - field element decomposition:
+            //      - rho = e_1 + (2^4) g + (2^254) h_0
+            //      - psi = h_1 + (2^9) i + (2^249) j_0 + (2^254) j_1
+            //   - canonicity:
+            //      - b_1 = 0 => b_0 = 0
+            //                && z13_a = 0
+            //                && a_prime_decomposition = 0
+            //      - d_0 = 0 => z13_c = 0
+            //                && b3_c_prime_decomposition = 0
+            //      - h_0 = 0 => z13_f = 0
+            //                && e1_f_prime_decomposition = 0
+            //      - j_1 = 0 => j_0 = 0
+            //                && g1_g2_prime_decomposition = 0
+
+            let q_canon_2 = meta.query_selector(config.q_canon_2);
+
+            // Offset prev
+            let e_1 = meta.query_advice(config.advices[5], Rotation::prev());
+            let f = meta.query_advice(config.advices[6], Rotation::prev());
+            let g_1 = meta.query_advice(config.advices[8], Rotation::prev());
```

Comment:
We can avoid these rotations entirely if we swap these three cells with the three cells at `Rotation::cur()` that are unused in this gate.